### PR TITLE
chore(deps): Update aws-sdks to 0.23 and aws supporting crates to 0.53

### PR DESCRIPTION
**Pull request by spencergilbert**
Alternative to closing #16016

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -168,21 +168,22 @@ metrics = "0.20.1"
...
-aws-smithy-http = { version = "0.51.0", default-features = false, features = ["event-stream"], optional = true }
-aws-smithy-http-tower = { version = "0.54.1", default-features = false, optional = true }
-aws-smithy-types = { version = "0.51.0", default-features = false, optional = true }
+aws-config = { version = "0.53.0", default-features = false, features = ["rustls"], optional = true }
+aws-credential-types = { version = "0.53.0", default-features = false, features = ["hardcoded-credentials"], optional = true }
+aws-smithy-async = { version = "0.53.1", default-features = false, optional = true }
+aws-smithy-client = { version = "0.53.1", default-features = false, features = ["client-hyper"], optional = true}
+aws-smithy-http = { version = "0.53.1", default-features = false, features = ["event-stream"], optional = true }
+aws-smithy-http-tower = { version = "0.53.1", default-features = false, optional = true }
+aws-smithy-types = { version = "0.53.1", default-features = false, optional = true }
```
Comment (spencergilbert) [inline Cargo.toml:186]: This isn't quite latest, but the next bump pulled in some changes to the credentials cache I wasn't super happy/comfortable with.

File: src/aws/region.rs [hunk 2]
```diff
@@ -36,9 +32,9 @@ impl RegionOrEndpoint {
         }
     }
 
-    pub fn endpoint(&self) -> crate::Result<Option<Endpoint>> {
-        let uri = self.endpoint.as_deref().map(Uri::from_str).transpose()?;
-        Ok(uri.map(Endpoint::immutable))
+    pub fn endpoint(&self) -> crate::Result<Option<String>> {
+        let endpoint = self.endpoint.clone();
+        Ok(endpoint)
```
Comment (spencergilbert) [inline src/aws/region.rs:37]: Doesn't strictly need to still be a `Result` but I figured we'd want to reintroduce some endpoint validation at some point and this minimized unnecessary changes in the codebase.

File: src/sinks/aws_cloudwatch_logs/retry.rs [hunk 3]
```diff
@@ -36,23 +36,26 @@ impl<T: Send + Sync + 'static> RetryLogic for CloudwatchRetryLogic<T> {
     fn is_retriable_error(&self, error: &Self::Error) -> bool {
```
Comment (spencergilbert) [inline src/sinks/aws_cloudwatch_logs/retry.rs:36]: I suspect we don't need these specific matchers and can delegate to the shared `is_retriable_error` but didn't want to pull that into this PR.

File: src/common/s3.rs [hunk 4]
```diff
@@ -12,6 +12,10 @@ impl ClientBuilder for S3ClientBuilder {
     }
 
     fn build(client: aws_smithy_client::Client, config: &aws_types::SdkConfig) -> Self::Client {
-        aws_sdk_s3::client::Client::with_config(client, config.into())
+        let config = aws_sdk_s3::config::Builder::from(config)
+            .force_path_style(true)
+            .build();
+
+        aws_sdk_s3::client::Client::with_config(client, config)
```
Comment (spencergilbert) [inline src/common/s3.rs:19]: This forces us into path style requests which localstack is more amenable to, theoretically [they support it](https://docs.localstack.cloud/user-guide/aws/s3/#getting-started) but I couldn't get it working in our integration test setup.

Virtual hosting is preferred, but the deprecation of path style has been delayed.

If we wanted to play around some more to get it working with localstack I would be happy to pair.

**Review by bruceg — APPROVED** [review]
I have some suggestions below for cleanups, but nothing blocking.

File: src/aws/mod.rs [hunk 5]
```diff
@@ -245,16 +263,31 @@ where
...
+                    SdkBody::from_dyn(BoxBody::new(body))
+                })
+            };
+
+            let req = Request::from_parts(http::Request::from_parts(parts, body), properties);
+
+            (req, Some(shared_bytes_sent))
+        } else {
+            (req, None)
+        };
```
[the remark this answers is not in the document]
Comment (spencergilbert) [reply src/aws/mod.rs:290]: 🚀

File: src/aws/mod.rs [hunk 6]
```diff
@@ -283,69 +316,56 @@ where
...
-            .store(self.bytes_sent, Ordering::Release);
+impl Body for MeasuredBody {
+    type Data = Bytes;
+    type Error = Box<dyn Error + Send + Sync>;
+
+    fn poll_data(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Result<Self::Data, Self::Error>>> {
+        self.poll_inner(cx)
```
[the remark this answers is not in the document]
Comment (spencergilbert) [reply src/aws/mod.rs:362]: I'm not entirely sure what you mean here?

[the remark this answers is not in the document]
Comment (bruceg) [reply src/aws/mod.rs:362]: The trait implementation just calls the `poll_inner` function which could be moved into the caller. I.e. move lines 340-350 into this function and drop `fn poll_inner`

File: src/aws/mod.rs [hunk 7]
```diff
@@ -41,41 +43,57 @@ pub fn is_retriable_error<T>(error: &SdkError<T>) -> bool {
...
+        SdkError::ResponseError(err) => {
+            let raw = err.raw();
+
+            check_response(raw)
+        }
+        SdkError::ServiceError(err) => {
+            let raw = err.raw();
+
+            check_response(raw)
+        }
```
[the remark this answers is not in the document]
Comment (spencergilbert) [reply src/aws/mod.rs]: Resolved in [40124b0](https://github.com/vectordotdev/vector/pull/16365/commits/40124b0c83d8a7910f20b0ffdbd6ce71731ab4ad)

File: src/aws/mod.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (spencergilbert) [reply src/aws/mod.rs]: Moved in c906f7bfb

**Review by neuronull — COMMENTED** [review]
LGTM-

I'm not super knowledgable about the aws sources, so not much help there.
Didn't notice anything while scanning through the changes.
I defer to CI and others :)

File: Cargo.toml [hunk 8]
```diff
@@ -168,21 +168,22 @@ metrics = "0.20.1"
...
 # AWS - Official SDK
-aws-sdk-s3 = { version = "0.21.0", default-features = false, features = ["native-tls"], optional = true }
-aws-sdk-sqs = { version = "0.21.0", default-features = false, features = ["native-tls"], optional = true }
-aws-sdk-cloudwatch = { version = "0.21.0", default-features = false, features = ["native-tls"], optional = true }
-aws-sdk-cloudwatchlogs = { version = "0.21.0", default-features = false, features = ["native-tls"], optional = true }
-aws-sdk-elasticsearch = {version = "0.21.0", default-features = false, features = ["native-tls"], optional = true }
-aws-sdk-firehose = { version = "0.21.0", default-features = false, features = ["native-tls"], optional = true }
-aws-sdk-kinesis = { version = "0.21.0", default-features = false, features = ["native-tls"], optional = true }
-aws-types = { version = "0.51.0", default-features = false, features = ["hardcoded-credentials"], optional = true }
+aws-sdk-s3 = { version = "0.23.0", default-features = false, features = ["rustls"], optional = true }
```
Comment (jszwedko) [inline Cargo.toml:171]: Just a note to swap these back to `native-tls`.

Comment (spencergilbert) [reply Cargo.toml:171]: Doh!

Comment (spencergilbert) [reply Cargo.toml]: Fixed in [0a5ad8d](https://github.com/vectordotdev/vector/pull/16365/commits/0a5ad8dc4d156db6fb5027d4ca9db9451a6d90b0)

**Review by jszwedko — APPROVED** [review]
